torch/_higher_order_ops/utils.py

```diff
@@ -76,15 +76,29 @@ def graph_with_interpreter(*args):
     return maybe_interpreted_fn
 
 
+_is_compiling_hoo_flag: bool = False
+
+
+def _is_compiling_hoo() -> bool:
```

Comment:
hmm did we need this due to circular imports?

Reply:
Actually, this is a part of my change that I am not 100% sure about. I've noticed a recurring pattern in the HOO code  like:

```python
if torch._dynamo.is_compiling():
    return some_higher_order_op()

<...>
return torch.compile(some_higher_order_op, ...)
```

Since in this PR I made sure to set `is_compiling` flag to `True`, the code inside `if`-block started to get executed instead of going inside `torch.compile()`, so I thought that probably a separate flag should be created for HOO case.

cc @tugsbayasgalan, @suo

Reply:
> torch._dynamo.is_compiling() returning True in this case is the existing behavior right?

Yeah, previously it would return `False` for non-strict mode, thus we would never go inside the `if`-block. Now for non-strict mode it returns `True`, so we get inside the `if`-block and fail with `FunctionalTensorMode`-related error.

I don't fully understand this `if`-block, tried to find unit tests that would cover it, but didn't find anything, therefore any opinion from HOO experts would be appreciated.

Reply:
cc @zou3519 do you know about this?

Reply:
n00q, what's does the "HOO" mean? I didn't find any clue related to it, thanks

Reply:
I don't think a separate `is_compiling_hoo()` makes sense.

Looking at the `cond` example, I believe the desired behavior is for us to go down the `True` path in non-strict mode, where we will call the higher-order op and then drop into the ProxyTensorMode implementation.

But it appears you have tripped a latent issue, which is that `torch.cond` and friends don't work with non-strict mode. So we probably should address that directly (not in this PR).

For now, let's just disable any non-strict tests that fail with the new behavior. This not a regression, since they were silently broken anyway. 

@ydwu4, please double-check my reasoning here + can you follow-up on the control flow support in non-strict?

Reply:
Okay, that means I was wrong, and that the desired behavior (at least today) is that we go down the `False` branch in non-strict mode. Which means some version of this is necessary.

@khabinov: my recommendation is that instead of a special `is_compiling_hoo`, we make a more general-purpose `is_dynamo_compiling` flag, which is true only if we are in dynamo (basically the old behavior of `torch._dynamo.is_compiling`), and we use it here specifically. That would communicate the intention of this code better.